lib/WTSI/NPG/OM/BioNano/Publisher.pm

```diff
@@ -0,0 +1,270 @@
+use File::Spec;
+use UUID;
+use URI;
+
+use WTSI::NPG::iRODS;
+use WTSI::NPG::iRODS::Collection;
+use WTSI::NPG::iRODS::DataObject;
+use WTSI::NPG::iRODS::Metadata;
+use WTSI::NPG::HTS::Publisher;
+use WTSI::NPG::OM::BioNano::ResultSet;
+
+# FIXME Move/refactor WTSI::NPG::HTS::Publisher to reflect use outside of
+# HTS. Maybe consolidate with WTSI::NPG::Publisher in wtsi-npg/genotyping.
+
+our $VERSION = '';
+
+with qw/WTSI::DNAP::Utilities::Loggable
+        WTSI::NPG::Accountable
+        WTSI::NPG::OM::BioNano::Annotator/;
+
+has 'accountee_uid' =>
+  (is       => 'ro',
+   isa      => 'Maybe[Str]',
+   required => 0,
+   lazy     => 1,
+   default  => undef,
+   documentation => 'UID (eg. username) supplied to HTS::Publisher for '.
+       'constructing the user URI in creation metadata. Optional, assigned '.
+       'a default in WTSI::NPG::Accountable if not given.',
+);
+
+has 'affiliation_uri' =>
+  (is       => 'ro',
+   isa      => 'Maybe[URI]',
+   required => 0,
+   lazy     => 1,
+   default  => undef,
+   documentation => 'Affiliation URI supplied to HTS::Publisher. Optional, '.
+       'assigned a default in WTSI::NPG::Accountable if not given.',
+   );
+
+has 'irods' =>
+  (is       => 'ro',
+   isa      => 'WTSI::NPG::iRODS',
+   required => 1,
+   default  => sub {
+     return WTSI::NPG::iRODS->new;
+   });
+
+has 'resultset' =>
+  (is       => 'ro',
+   isa      => 'WTSI::NPG::OM::BioNano::ResultSet',
+   required => 1);
+
+has 'uuid' =>
+  (is       => 'ro',
+   isa      => 'Str',
+   required => 1,
+   lazy     => 1,
+   builder  => '_build_uuid',
+   documentation => 'UUID generated for the publication to iRODS');
+
+sub BUILD {
+  my ($self) = @_;
+
+  # Make our irods handle use our logger by default
+  $self->irods->logger($self->logger);
+
+  return 1;
+}
+
+our @BNX_SUFFIXES = qw[bnx];
+
+=head2 publish
+
+  Arg [1]    : [Str] iRODS path that will be the root destination for
+               publication. BioNano will be published to a subcollection,
+               with a hashed path based on the md5 checksum of the
+               Molecules.bnx file.
+  Arg [2]    : [DateTime] Timestamp for the time of publication. Optional,
+               defaults to the present time.
+
+  Example    : $export->publish('/foo')
+  Description: Publish the BioNano ResultSet to an iRODS path.
+  Returntype : True
+
+=cut
+
+sub publish {
+    my ($self, $publish_dest, $timestamp) = @_;
+    # generate a hashed path for publication
+    my $hash_path =
+        $self->irods->hash_path($self->resultset->bnx_path,
+                                $self->resultset->bnx_file->md5sum);
+    $self->debug(q{Found hashed path '}, $hash_path, q{' from checksum '},
+                 $self->resultset->bnx_file->md5sum, q{'});
+    if (! defined $timestamp) {
+        $timestamp = DateTime->now();
+    }
+    if (! File::Spec->file_name_is_absolute($publish_dest)) {
+        $publish_dest = File::Spec->catdir($self->irods->working_collection,
```

Comment:
If someone provides a relative iRODS path as publish_dest, an exception should be raised.